stdlib/regex.gr

```diff
@@ -0,0 +1,2995 @@
+        let (min1, max1, maxL1) = loop(re)
+        if (min1 == 0) {
+          thrownError = Some(MightBeEmpty)
+          (0, 0, 0)
+        } else {
+          mustSizes = mergeDependsSizes(oldMustSizes, mustSizes)
+          dependsSizes = mergeDependsSizes(oldDependsSizes, dependsSizes)
+          let repeatMax = match(repeatMax) {
+            None => Float32.toNumber(Float32.infinity),
+            Some(n) => n
+          }
+          (min1 * repeatMin, max1 * repeatMax, maxL1)
+        }
+      },
+      REMaybe(re, nonGreedy) => {
+        let (min1, max1, maxL1) = loop(re)
+        (0, max1, maxL1)
+      },
+      REConditional(reTest, reTrue, reFalse, _, _, _) => {
+        let (min1, max1, maxL1) = loop(reTest)
+        let (min2, max2, maxL2) = loop(reTrue)
+        let (min3, max3, maxL3) = Option.mapWithDefault(loop, (0, 0, 0), reFalse)
+        (min(min2, min3), max(max2, max3), max(max(maxL1, maxL2), maxL3))
+      },
+      RELookahead(re, _, _, _) => {
+        let (min1, max1, maxL1) = loop(re)
+        (0, 0, maxL1)
+      },
+      RELookbehind(re, _, lbMin, lbMax, _, _) => {
+        let (min1, max1, maxL1) = loop(re)
+        if (max1 == Float32.toNumber(Float32.infinity)) {
+          thrownError = Some(DoesNotMatchBounded)
+          (0, 0, 0)
+        } else {
+          lbMin := min1
+          lbMax := max1
+          (0, 0, max(max1, maxL1))
+        }
+      },
+      RECut(re, _, _, _) => {
+        loop(re)
+      },
+      REReference(n, _) => {
+        if (n > numGroups) {
+          thrownError = Some(BackreferenceTooBig)
+          (0, 0, 0)
+        } else {
+          match(Map.get(n, groupSizes)) {
+            Some(minSize) => (minSize, Float32.toNumber(Float32.infinity), 0),
+            None => {
+              Map.set(n - 1, true, dependsSizes)
+              (1, Float32.toNumber(Float32.infinity), 0)
+            }
+          }
+        }
+      },
+      REUnicodeCategories(_, _) => (1, 4, 0)
+    }
+  }
+  let (minLen, maxLen, maxLookbehind) = loop(re)
+  Map.forEach((k, _) => {
+    match(Map.get(k, groupSizes)) {
+      None => void,
+      Some(sz) => {
+        if (sz <= 0) {
+          thrownError = Some(MightBeEmpty)
+        }
+      }
+    }
+  }, mustSizes)
+  match(thrownError) {
+    Some(MightBeEmpty) => Err("`*`, `+`, or `{...}` operand could be empty"),
+    Some(DoesNotMatchBounded) => Err("lookbehind pattern does not match a bounded length"),
+    Some(BackreferenceTooBig) => Err("backreference number is larger than the highest-numbered cluster"),
+    Some(InternalError(re)) => Err("regex validate: Internal error: " ++ toString(re)),
+    None => Ok(maxLookbehind)
+  }
+}
+
+
+/*
+
+=========================
+REGEX MATCHER COMPILATION
+=========================
+
+*/
+
+
+record MatchBuf {
+  matchInput: String,
+  matchInputExploded: Array<Char>,
+}
+
+let makeMatchBuffer = (s) => {
+  {
+    matchInput: s,
+    matchInputExploded: String.explode(s),
```

Comment:
This is still my only major worry, but I don't think it needs to be refactored right now (especially since it'd be a pretty big refactor).

Reply:
@ospencer I apologize, can you remind me what your worry is here?

Reply:
It's that this is inefficient and may be problematic for larger string sizes since the memory requirement is 28x character length of string (24 bytes for each Char and 4 bytes for each array slot) + computation time. https://github.com/grain-lang/grain/issues/769 is possibly a fix here? I'm not sure what Racket does, but maybe it treats multiple-byte characters as just a sequence to match against. Gives you the byte offsets instead of the character offsets though—not sure how expensive it would be to convert them. But then there'd be no memory overhead and no conversion cost! Otherwise, I think converting to UTF-32 instead of exploding might be better. That'd be 4x character length of string with constant time indexing. 

Regardless, we've got some options for what we can do here (that don't need to happen right now). If you don't think #769 is a fix, could you make an issue for it?

Reply:
#901